spec.bs

```diff
@@ -8159,43 +8321,154 @@ To <dfn>batch or fetch trusted bidding signals</dfn> given a [=trusted bidding s
-      |ig|'s [=interest group/max trusted bidding signals url length=].
+    1. Let |keyValueInterestGroups| be |trustedBiddingSignalsBatcher|'s [=trusted bidding signals
+      batcher/key value interest groups=].
+    1. Let |key| be [=tuple=] of (|ig|'s [=interest group/joining origin=], |ig|'s [=interest group/owner=]).
+    1. If |keyValueInterestGroups| [=map/contains=] |key|, [=set/append=] |ig| to
+      |keyValueInterestGroups|[|key|].
+    1. Otherwise, set |keyValueInterestGroups|[|key|] to a [=set=] « |ig| ».
+
+</div>
+
+<div algorithm>
+
+To <dfn>build trusted key value bidding signals request body</dfn> given a [=set=] |interestGroups|,
+an {{unsigned short}}-or-null |experimentGroupId|, an [=origin=] |topLevelOrigin|, a [=string=]
+|slotSizeQueryParam|, an [=origin=] |coordinator|, and an [=origin=] |owner|:
+
+1. Let |compressionGroups| be an empty [=list=], whose [=list/items=] are [=compression groups=].
+1. Let |compressionGroupMap| be an empty [=map=], where the [=map/keys=] are integers as
+  [=compression group/compression group id=], and the [=map/values=] are [=maps=] with integers as [=map/keys=]
+  for [=partition/id=], and [=partitions=] as their [=map/values=].
+1. Let |compressionIdMap| be an empty [=map=], whose [=map/keys=] are [=origins=] and [=map/values=] are integers.
+1. Let |interestGroupIdMap| be an empty [=map=], whose [=map/keys=] are [=strings=] and [=map/values=] are [=tuples=]
+  of (interger, integer).
+1. Let |slotSizeParams| be the result of [=strictly splitting=] |slotSizeQueryParam| on U+003D (=).
+1. Let |nextCompressionGroupId| be 0.
+1. [=list/For each=] |group| of |interestGroups|:
+  1. Set |joiningOrigin| to |group|'s [=interest group/joining origin=].
+  1. Let |partitionId| be an integer.
+  1. If |compressionIdMap| does not [=map/contain=] |joiningOrigin|:
+    1. [=map/Set=] |compressionIdMap|[|joiningOrigin|] to |nextCompressionGroupId|.
+    1. Increment |nextCompressionGroupId| by 1.
+  1. Let |compressionGroupId| be |compressionIdMap|[|joiningOrigin|].
+  1. If |compressionGroupMap| does not [=map/contain=] |compressionGroupId|:
+    1. Let |compressionGroupMap|[|compressionGroupId|] be an empty [=map=], whose [=map/keys=] are integers
+      and [=map/values=] are [=maps=].
+  1. Set |executionMode| to |group|'s [=interest group/execution mode=].
+  1. If |executionMode| equal to "`group-by-origin`", set |partitionId| to 0.
+  1. Otherwise:
+    1. If |compressionGroupMap|[|compressionGroupId|] [=map/contains=] 0, set |partitionId| to
+      [=map/size=] of |compressionGroupMap|[|compressionGroupId|].
+    1. Otherwise, set |partitionId| to sum [=list/size=] of |compressionGroupMap|[|joiningOrigin|]
+      and 1.
+  1. Set |interestGroupIdMap|[|group|'s [=interest group/name=]] to [=tuple=] of |compressionGroupId|
+    and |partitionId|.
+  1. If |compressionGroupMap|[|compressionGroupId|] does not [=map/contain=] |partitionId|:
+    1. Let |partition| be an empty [=partition=].

```

Comment:
same comment as in the scoring algorithm, it'll be much clearer to use something like https://github.com/WICG/turtledove/blob/main/spec.bs#L890